src/main/ApplicationImpl.cpp

```diff
@@ -475,6 +480,59 @@ ApplicationImpl::reportInfo()
     LOG_INFO(DEFAULT_LOG, "info -> {}", getJsonInfo().toStyledString());
 }
 
+std::shared_ptr<BasicWork>
+ApplicationImpl::scheduleSelfCheck(bool waitUntilNextCheckpoint)
+{
+    // We store any self-check that gets scheduled so that a second call to
+    // schedule a self-check while one is running returns the running one.
+    if (*mRunningSelfCheck)
+    {
+        return *mRunningSelfCheck;
+    }
+    auto& ws = getWorkScheduler();
+    auto& lm = getLedgerManager();
+    auto& ham = getHistoryArchiveManager();
+
+    LedgerHeaderHistoryEntry lhhe = lm.getLastClosedLedgerHeader();
+
+    std::vector<std::shared_ptr<BasicWork>> seq;
+    seq.emplace_back(ham.getHistoryArchiveReportWork());
+
+    auto checkLedger = ham.getCheckLedgerHeaderWork(lhhe);
+    if (waitUntilNextCheckpoint)
+    {
+        // Delay until a second full checkpoint-period after the next checkpoint
+        // publication. The captured lhhe should usually be published by then.
+        auto& hm = getHistoryManager();
+        auto targetLedger =
+            hm.firstLedgerAfterCheckpointContaining(lhhe.header.ledgerSeq);
+        targetLedger = hm.firstLedgerAfterCheckpointContaining(targetLedger);
+        std::weak_ptr<LedgerManager> weakLm(mLedgerManager);
+        auto cond = [weakLm, targetLedger]() -> bool {
+            auto lm = weakLm.lock();
+            return lm && lm->getLastClosedLedgerNum() > targetLedger;
+        };
+        seq.emplace_back(std::make_shared<ConditionalWork>(
+            *this, "wait-for-lcl", cond, checkLedger,
+            std::chrono::seconds(10)));
+    }
+    else
+    {
+        seq.emplace_back(checkLedger);
+    }
+
+    // When a self-check completes, it resets the cell holding the reference to
+    // itself, allowing another self-check to start.
+    auto cell = mRunningSelfCheck;
```

Comment:
I don't think we need the double pointer and the (somewhat complicated) resetting mechanism here. Can we store a single pointer to the work sequence and let `scheduleWork` decide what to do based on its state? I.e., every time this function is called, check if we have the work set. If it's done, reset it and spin up a new one. Otherwise, return the existing pointer.

Reply:
Fixed this with the retained weak_ptr that @MonsieurNicolas suggested.